store/src/java/com/zimbra/cs/imap/ImapServerListener.java

```diff
@@ -346,10 +346,11 @@ public String getWSId() {
         return wsID;
     }
 
-    private synchronized void cancelPendingRequest() {
-        ZimbraLog.imap.debug("Canceling pending AdminWaitSetRequest for waitset %s. Sequence %s", wsID, lastSequence.toString());
-        if(pendingRequest != null && !(pendingRequest.isCancelled() || pendingRequest.isDone())) {
+    private void cancelPendingRequest() {
+        if (pendingRequest != null && !(pendingRequest.isCancelled() || pendingRequest.isDone())) {
+            ZimbraLog.imap.debug("Canceling pending AdminWaitSetRequest for waitset %s. Sequence %s", wsID, lastSequence.toString());
             pendingRequest.cancel(true);
+            pendingRequest = null;
```

Comment:
Is it possible for cancelPendingRequest to be called from more than one thread?  If it is, there are possibilities for NPEs here.  Having said that, can probably fix that by assigning it to another variable or guard it with an atomic perhaps.

Reply:
The only time i EVER see it used is when it is assigned and that usage IS NOT protected.
Does it need to be?

Reply:
Well the lock on this class itself is called from more than one thread.  This is the lock that was deadlocked in the crash dump.  

Would you be happy with the 'synchronized' removal if I remove the assignment?
This would prevent any NPEs since this function is the only thing that actually operates on the 'pendingRequest' variable directly.

Reply:
@grishick added the synchronized keyword in change  14f483194e72b7edf2d5557c1bca4909519ea67f
Perhaps he would be best to explain the impact?

Reply:
I see how this synchronization is a problem. The reason for having the keyword there is to avoid trying to double-cancel the same Future. If you are removing the keyword, it probably makes sense to surround pendingRequest.cancel(true); with try-catch block and swallow the exception when Future has already been cancelled.

Reply:
It appears based on the documentation that it simply returns 'false' in the case that it is already canceled.  

https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Future.html#cancel(boolean)

Does reality not match the documentation?

Reply:
I'll test it.